### ISSUE TITLE
Emscripten emits signed values when return type is uint8_t

### ISSUE THREAD

**Issue by awalgarg** [emscripten-core/emscripten#5189]
The following C code:

```
#include<stdint.h>
uint8_t f() { return *((uint8_t*)(1)); }
```

When compiled with:

```
emcc foo.c -Os -s WASM=1 -s SIDE_MODULE=1 -o a.wasm
```

and dumping the wasm ast with wasm2wast, shows the following:

```
  (func (;0;) (type 0) (result i32)
    i32.const 1
    i32.load8_s)
```

Shouldn't `load8_s` should be `load8_u` to produce an unsigned value? Interestingly, switching the return type in the C code to `int32_t` for the function signature (in)correctly produces `load8_u`.

Apologies if this was already asked/reported.

cc @copy

**Comment by kripken** [comment]
I believe the underlying reason for this is that at the lowest level (CPUs, LLVM, ABIs) there is no signed vs unsigned 8 bit value. There is just "8 bits". That's why LLVM emits this, for both int and uint:
````
define zeroext i8 @f() local_unnamed_addr #0 {
````
All it has is `i8` in both cases. Where the sign can come into play is when the value is used. Wasm, like LLVM, similarly doesn't have signed vs unsigned types.

But where this gets confusing is that wasm doesn't have 8-bit types, it has just 32 bits (and 64). As a result, we need to extend the 1-byte load in either a signed or unsigned way. We can actually do either, as it is just where the value is used that it matters, and so the binaryen optimizer will look at the uses and find the best way. But here, there are no uses aside from returning the value. So it's arbitrary.

tl;dr, wasm has no 8 bit type, so when you return a 32-bit type, only the 8 low bits matter. The upper bits are ignored, so they could make it look like the 32-bit value is signed or unsigned. In other words, when you receive this value on the other side, you must look at just those low bits, and interpret them as signed or not as appropriate.
